### PR TITLE
feat(storage): Add cname support for V4 signature

### LINKED ISSUE #11: [Storage] V4 Signature Signed URLs With Bucket-Bound Hostnames
Tracking issue for work defined in companion document.

[New Client Request - Support for Bucket-Bound Hostnames With GCS Signed URLs](https://docs.google.com/document/d/1Da8LQEmc1OKRECxXqdivQa4tcTEImlti_fVYzoh4reI)

### PR DESCRIPTION
**Pull request by HemangChothani**
Fixes #11

### CODE REVIEW

File: google/cloud/storage/blob.py [hunk 1]
```diff
@@ -460,6 +462,18 @@ def generate_signed_url(
...
             virtual hostname, e.g., '<bucket-name>.storage.googleapis.com'.
 
+        :type bucket_bound_host_name: str
+        :param bucket_bound_host_name:
+            (Optional) If pass, then construct the URL relative to the bucket-bound hostname
+            as a CNAME. Value cane be a bare or with scheme, e.g., 'example.com ' or http://example.com.
+            See: https://cloud.google.com/storage/docs/request-endpoints#cname
+
+        :type host_name_scheme: str
+        :param host_name_scheme:
```
Comment (frankyn) [inline google/cloud/storage/blob.py]: prefer scheme here.

File: google/cloud/storage/bucket.py [hunk 2]
```diff
@@ -2373,6 +2375,8 @@ def generate_signed_url(
         amount of time, you can use this method to generate a URL that
         is only valid within a certain time period.
 
+        If ``cname`` is set as an argument of :attr:`api_access_endpoint`, ``https`` works only if using a ``CDN``.
```
Comment (frankyn) [inline google/cloud/storage/bucket.py]: cname no longer used in method signature should be updated.

File: google/cloud/storage/bucket.py [hunk 3]
```diff
@@ -2440,10 +2456,19 @@ def generate_signed_url(
             api_access_endpoint = "https://{bucket_name}.storage.googleapis.com".format(
                 bucket_name=self.name
             )
-            resource = "/"
+        elif bucket_bound_host_name:
+            if ":" in bucket_bound_host_name:
+                api_access_endpoint = bucket_bound_host_name
+            else:
+                api_access_endpoint = "{scheme}://{cname}".format(
```
Comment (frankyn) [inline google/cloud/storage/bucket.py]: use `bucket_bound_hostname` to not go back and forth between `cname`

File: google/cloud/storage/bucket.py [hunk 4]
```diff
@@ -2373,6 +2375,8 @@ def generate_signed_url(
         amount of time, you can use this method to generate a URL that
         is only valid within a certain time period.
 
+        If ``cname`` is set as an argument of :attr:`api_access_endpoint`, ``https`` works only if using a ``CDN``.
+
```
Comment (frankyn) [inline google/cloud/storage/bucket.py]: Add an inline example of using `bucket_bound_hostname` and `scheme`

File: google/cloud/storage/bucket.py [hunk 5]
```diff
@@ -2355,6 +2355,8 @@ def generate_signed_url(
         credentials=None,
         version=None,
         virtual_hosted_style=False,
+        bucket_bound_host_name=None,
```
Comment (frankyn) [inline google/cloud/storage/bucket.py]: use: `bucket_bound_hostname`

File: tests/unit/test__signing.py [hunk 6]
```diff
@@ -744,14 +746,31 @@ def test_conformance_client(test_data):
 
 @pytest.mark.parametrize("test_data", _BUCKET_TESTS)
 def test_conformance_bucket(test_data):
-    resource = "/{}".format(test_data["bucket"])
-    _run_conformance_test(resource, test_data)
+    if "urlStyle" in test_data and test_data["urlStyle"] == "BUCKET_BOUND_DOMAIN":
```
Comment (frankyn) [inline tests/unit/test__signing.py]: Please update to latest release of conformance tests this value was updated to `BUCKET_BOUND_HOSTNAME`. Last nit.

**Comment by frankyn** [comment]
Friendly ping, @HemangChothani

**Comment by HemangChothani** [comment]
@frankyn   Do we need to add conformance tests for `virtual_hosted_style`?

**Comment by frankyn** [comment]
Good, catch, that was not implemented in #58 could you pick that up?

**Comment by HemangChothani** [comment]
@frankyn Yes sure, thanks.

**Comment by frankyn** [comment]
@tseaver was working on this in https://github.com/googleapis/python-storage/issues/57

**Pull request merged**